src/main/java/net/minecraftforge/common/ForgeHooks.java

```diff
@@ -1338,28 +1350,72 @@ public static void readAdditionalLevelSaveData(LevelStorageSource.LevelDirectory
                 {
                     continue;
                 }
+
                 String modVersion = mod.getString("ModVersion");
-                ModList.get().getModContainerById(modId).ifPresentOrElse(container -> {
+                final var previousVersion = new DefaultArtifactVersion(modVersion);
+                ModList.get().getModContainerById(modId).ifPresentOrElse(container ->
+                {
                     final var loadingVersion = container.getModInfo().getVersion();
-                    final var previousVersion = new DefaultArtifactVersion(modVersion);
                     if (!loadingVersion.equals(previousVersion))
                     {
                         // Enqueue mismatched versions for bulk event
                         mismatchedVersions.put(modId, previousVersion);
                     }
-                }, () -> LOGGER.error(WORLDPERSISTENCE,"This world was saved with mod {} which appears to be missing, things may not work well", modId));
+                }, () -> missingVersions.put(modId, previousVersion));
             }
 
-            final var mismatchEvent = new ModVersionsMismatchEvent(levelDirectory, mismatchedVersions);
+            final var mismatchEvent = new ModMismatchEvent(levelDirectory, mismatchedVersions, missingVersions);
             ModLoader.get().postEvent(mismatchEvent);
 
+            StringBuilder resolved = new StringBuilder("The following mods have version differences that were marked resolved:");
+            StringBuilder unresolved = new StringBuilder("The following mods have version differences that were not resolved:");
+
+            // For mods that were marked resolved, log the version resolution and the mod that resolved the mismatch
+            mismatchEvent.getResolved().forEachOrdered((res) ->
+            {
+                final var modid = res.modid();
+                final var diff = res.versionDifference();
+                if(res.wasSelfResolved()) {
+                    resolved.append(System.lineSeparator())
+                            .append(diff.isMissing() ?
+                                    "%s (version %s -> MISSING, self-resolved)".formatted(modid, diff.oldVersion()) :
+                                    "%s (version %s -> %s, self-resolved)".formatted(modid, diff.oldVersion(), diff.newVersion())
```

Comment:
I'm not going to request this be changed now, but it might be better readability-wise if the ternary operators are placed at the beginning of the continuation lines, rather than at the end.
```suggestion
                            .append(diff.isMissing()
                                    ? "%s (version %s -> MISSING, self-resolved)".formatted(modid, diff.oldVersion())
                                    : "%s (version %s -> %s, self-resolved)".formatted(modid, diff.oldVersion(), diff.newVersion())
```